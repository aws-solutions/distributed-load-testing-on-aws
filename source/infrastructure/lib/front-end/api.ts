// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Code, Function as LambdaFunction, Runtime } from "aws-cdk-lib/aws-lambda";
import { Aws, ArnFormat, CfnResource, Duration, RemovalPolicy, Stack } from "aws-cdk-lib";
import { IBucket } from "aws-cdk-lib/aws-s3";
import { LogGroup, RetentionDays } from "aws-cdk-lib/aws-logs";
import { Effect, Policy, PolicyDocument, PolicyStatement, Role, ServicePrincipal } from "aws-cdk-lib/aws-iam";
import {
  AccessLogFormat,
  AuthorizationType,
  CfnAccount,
  ContentHandling,
  Deployment,
  EndpointType,
  Integration,
  IntegrationType,
  LogGroupLogDestination,
  MethodLoggingLevel,
  MethodOptions,
  PassthroughBehavior,
  RequestValidator,
  RestApi,
  Stage,
} from "aws-cdk-lib/aws-apigateway";
import { Construct } from "constructs";

/**
 * @interface DLTAPIProps
 * DLTAPI props
 */
export interface DLTAPIProps {
  readonly cloudWatchLogsPolicy: Policy;
  // ECS CloudWatch Log Group
  readonly ecsCloudWatchLogGroup: LogGroup;
  // ECS Task Execution Role ARN
  readonly ecsTaskExecutionRoleArn: string;
  // History DynamoDB table policy
  readonly historyDynamoDbPolicy: Policy;
  // History DynamoDB table
  readonly historyTable: string;
  // Test scenarios S3 bucket
  readonly scenariosBucketName: string;
  // Scenarios DynamoDB table policy
  readonly scenariosDynamoDbPolicy: Policy;
  // Test scenarios S3 bucket policy
  readonly scenariosS3Policy: Policy;
  // Test scenarios DynamoDB table
  readonly scenariosTableName: string;
  // Task canceler ARN
  readonly taskCancelerArn: string;
  // Task Canceler Invoke Policy
  readonly taskCancelerInvokePolicy: Policy;
  // Task Runner state function
  readonly taskRunnerStepFunctionsArn: string;

  /**
   * Solution config properties.
   * the metric URL endpoint, send anonymized usage, solution ID, version, source code bucket, and source code prefix
   */
  readonly metricsUrl: string;
  readonly sendAnonymizedUsage: string;
  readonly solutionId: string;
  readonly solutionVersion: string;
  readonly sourceCodeBucket: IBucket;
  readonly sourceCodePrefix: string;
  readonly uuid: string;
}

/**
 * Distributed Load Testing on AWS API construct
 */
export class DLTAPI extends Construct {
  apiId: string;
  apiEndpointPath: string;
  apiServicesLambdaRoleName: string;

  constructor(scope: Construct, id: string, props: DLTAPIProps) {
    super(scope, id);

    const taskArn = Stack.of(this).formatArn({
      service: "ecs",
      resource: "task",
      resourceName: "*",
      arnFormat: ArnFormat.SLASH_RESOURCE_NAME,
    });
    const taskDefArn = Stack.of(this).formatArn({ service: "ecs", resource: "task-definition/" });

    const dltApiServicesLambdaRole = new Role(this, "DLTAPIServicesLambdaRole", {
      assumedBy: new ServicePrincipal("lambda.amazonaws.com"),
      inlinePolicies: {
        DLTAPIServicesLambdaPolicy: new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["ecs:ListTasks"],
              resources: ["*"],
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["ecs:RunTask", "ecs:DescribeTasks"],
              resources: [taskArn, taskDefArn],
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["iam:PassRole"],
              resources: [props.ecsTaskExecutionRoleArn],
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["states:StartExecution"],
              resources: [props.taskRunnerStepFunctionsArn],
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["logs:DeleteMetricFilter"],
              resources: [props.ecsCloudWatchLogGroup.logGroupArn],
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["cloudwatch:DeleteDashboards"],
              resources: [
                Stack.of(this).formatArn({
                  service: "cloudwatch",
                  region: "",
                  resource: "dashboard",
                  resourceName: "EcsLoadTesting*",
                }),
              ],
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["cloudformation:ListExports"],
              resources: ["*"],
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: [
                "ecs:ListAccountSettings",
                "ecs:ListTasks",
                "ecs:ListClusters",
                "ecs:DescribeClusters",
                "ecs:DescribeTaskDefinition",
              ],
              resources: ["*"],
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["servicequotas:GetServiceQuota"],
              resources: ["*"],
            }),
          ],
        }),
      },
    });
    this.apiServicesLambdaRoleName = dltApiServicesLambdaRole.roleName;
    dltApiServicesLambdaRole.attachInlinePolicy(props.cloudWatchLogsPolicy);
    dltApiServicesLambdaRole.attachInlinePolicy(props.historyDynamoDbPolicy);
    dltApiServicesLambdaRole.attachInlinePolicy(props.scenariosDynamoDbPolicy);
    dltApiServicesLambdaRole.attachInlinePolicy(props.scenariosS3Policy);
    dltApiServicesLambdaRole.attachInlinePolicy(props.taskCancelerInvokePolicy);

    const ruleSchedArn = Stack.of(this).formatArn({ service: "events", resource: "rule", resourceName: "*Scheduled" });
    const ruleCreateArn = Stack.of(this).formatArn({ service: "events", resource: "rule", resourceName: "*Create" });
    const ruleListArn = Stack.of(this).formatArn({ service: "events", resource: "rule", resourceName: "*" });

    const lambdaApiEventsPolicy = new Policy(this, "LambdaApiEventsPolicy", {
      statements: [
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: ["events:PutTargets", "events:PutRule", "events:DeleteRule", "events:RemoveTargets"],
          resources: [ruleSchedArn, ruleCreateArn],
        }),
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: ["events:ListRules"],
          resources: [ruleListArn],
        }),
      ],
    });
    dltApiServicesLambdaRole.attachInlinePolicy(lambdaApiEventsPolicy);

    const apiLambdaRoleResource = dltApiServicesLambdaRole.node.defaultChild as CfnResource;
    apiLambdaRoleResource.addMetadata("cfn_nag", {
      rules_to_suppress: [
        {
          id: "W11",
          reason: "ecs:ListTasks and cloudformation:ListExports do not support resource level permissions",
        },
      ],
    });

    const dltApiServicesLambda = new LambdaFunction(this, "DLTAPIServicesLambda", {
      description: "API microservices for creating, updating, listing and deleting test scenarios",
      code: Code.fromBucket(props.sourceCodeBucket, `${props.sourceCodePrefix}/api-services.zip`),
      runtime: Runtime.NODEJS_16_X,
      handler: "index.handler",
      timeout: Duration.seconds(120),
      environment: {
        HISTORY_TABLE: props.historyTable,
        METRIC_URL: props.metricsUrl,
        SCENARIOS_BUCKET: props.scenariosBucketName,
        SCENARIOS_TABLE: props.scenariosTableName,
        SEND_METRIC: props.sendAnonymizedUsage,
        SOLUTION_ID: props.solutionId,
        STACK_ID: Aws.STACK_ID,
        STATE_MACHINE_ARN: props.taskRunnerStepFunctionsArn,
        TASK_CANCELER_ARN: props.taskCancelerArn,
        UUID: props.uuid,
        VERSION: props.solutionVersion,
      },
      role: dltApiServicesLambdaRole,
    });
    const apiLambdaResource = dltApiServicesLambda.node.defaultChild as CfnResource;
    apiLambdaResource.addMetadata("cfn_nag", {
      rules_to_suppress: [
        {
          id: "W58",
          reason: "CloudWatchLogsPolicy covers a permission to write CloudWatch logs.",
        },
        {
          id: "W89",
          reason: "VPC not needed for lambda",
        },
        {
          id: "W92",
          reason: "Does not run concurrent executions",
        },
      ],
    });

    const lambdaApiPermissionPolicy = new Policy(this, "LambdaApiPermissionPolicy", {
      statements: [
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: ["lambda:AddPermission", "lambda:RemovePermission"],
          resources: [dltApiServicesLambda.functionArn],
        }),
      ],
    });
    dltApiServicesLambdaRole.attachInlinePolicy(lambdaApiPermissionPolicy);

    const apiLogs = new LogGroup(this, "APILogs", {
      retention: RetentionDays.ONE_YEAR,
      removalPolicy: RemovalPolicy.RETAIN,
    });
    const apiLogsResource = apiLogs.node.defaultChild as CfnResource;
    apiLogsResource.addMetadata("cfn_nag", {
      rules_to_suppress: [
        {
          id: "W84",
          reason: "KMS encryption unnecessary for log group",
        },
      ],
    });

    const logsArn = Stack.of(this).formatArn({ service: "logs", resource: "*" });
    const apiLoggingRole = new Role(this, "APILoggingRole", {
      assumedBy: new ServicePrincipal("apigateway.amazonaws.com"),
      inlinePolicies: {
        apiLoggingPolicy: new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:DescribeLogGroups",
                "logs:DescribeLogStreams",
                "logs:PutLogEvents",
                "logs:GetLogEvents",
                "logs:FilterLogEvent",
              ],
              resources: [logsArn],
            }),
          ],
        }),
      },
    });

    const api = new RestApi(this, "DLTApi", {
      defaultCorsPreflightOptions: {
        allowOrigins: ["*"],
        allowHeaders: ["Authorization", "Content-Type", "X-Amz-Date", "X-Amz-Security-Token", "X-Api-Key"],
        allowMethods: ["DELETE", "GET", "HEAD", "OPTIONS", "PATCH", "POST", "PUT"],
        statusCode: 200,
      },
      deploy: true,
      deployOptions: {
        accessLogDestination: new LogGroupLogDestination(apiLogs),
        accessLogFormat: AccessLogFormat.jsonWithStandardFields(),
        loggingLevel: MethodLoggingLevel.INFO,
        stageName: "prod",
        tracingEnabled: true,
      },
      description: `Distributed Load Testing API - version ${props.solutionVersion}`,
      endpointTypes: [EndpointType.EDGE],
    });

    this.apiId = api.restApiId;
    this.apiEndpointPath = api.url.slice(0, -1);

    const apiAccountConfig = new CfnAccount(this, "ApiAccountConfig", {
      cloudWatchRoleArn: apiLoggingRole.roleArn,
    });
    apiAccountConfig.addDependency(api.node.defaultChild as CfnResource);
    const apiAllRequestValidator = new RequestValidator(this, "APIAllRequestValidator", {
      restApi: api,
      validateRequestBody: true,
      validateRequestParameters: true,
    });

    const apiDeployment = api.node.findChild("Deployment") as Deployment;
    const apiDeploymentResource = apiDeployment.node.defaultChild as CfnResource;
    apiDeploymentResource.addMetadata("cfn_nag", {
      rules_to_suppress: [
        {
          id: "W68",
          reason: "The solution does not require the usage plan.",
        },
      ],
    });

    const apiFindProdResource = api.node.findChild("DeploymentStage.prod") as Stage;
    const apiProdResource = apiFindProdResource.node.defaultChild as CfnResource;
    apiProdResource.addMetadata("cfn_nag", {
      rules_to_suppress: [
        {
          id: "W64",
          reason: "The solution does not require the usage plan.",
        },
      ],
    });

    const allIntegration = new Integration({
      type: IntegrationType.AWS_PROXY,
      integrationHttpMethod: "POST",
      options: {
        contentHandling: ContentHandling.CONVERT_TO_TEXT,
        integrationResponses: [{ statusCode: "200" }],
        passthroughBehavior: PassthroughBehavior.WHEN_NO_MATCH,
      },
      uri: `arn:${Aws.PARTITION}:apigateway:${Aws.REGION}:lambda:path/2015-03-31/functions/${dltApiServicesLambda.functionArn}/invocations`,
    });
    const allMethodOptions: MethodOptions = {
      authorizationType: AuthorizationType.IAM,
      methodResponses: [
        {
          statusCode: "200",
          responseModels: {
            "application/json": { modelId: "Empty" },
          },
        },
      ],
      requestValidator: apiAllRequestValidator,
    };

    /**
     * Test scenario API
     * /regions
     * /scenarios
     * /scenarios/{testId}
     * /tasks
     * /vCPUDetails
     */

    const regionsResource = api.root.addResource("regions");
    regionsResource.addMethod("ANY", allIntegration, allMethodOptions);

    const scenariosResource = api.root.addResource("scenarios");
    scenariosResource.addMethod("ANY", allIntegration, allMethodOptions);

    const testIds = scenariosResource.addResource("{testId}");
    testIds.addMethod("ANY", allIntegration, allMethodOptions);

    const tasksResource = api.root.addResource("tasks");
    tasksResource.addMethod("ANY", allIntegration, allMethodOptions);

    const vCPUDetails = api.root.addResource("vCPUDetails");
    vCPUDetails.addMethod("ANY", allIntegration, allMethodOptions);

    const invokeSourceArn = Stack.of(this).formatArn({
      service: "execute-api",
      resource: api.restApiId,
      resourceName: "*",
    });
    dltApiServicesLambda.addPermission("DLTApiInvokePermission", {
      action: "lambda:InvokeFunction",
      principal: new ServicePrincipal("apigateway.amazonaws.com"),
      sourceArn: invokeSourceArn,
    });
  }
}
