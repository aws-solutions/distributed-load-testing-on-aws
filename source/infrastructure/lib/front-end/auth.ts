// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import {
  UserPool,
  CfnUserPool,
  UserPoolClient,
  ClientAttributes,
  CfnIdentityPool,
  CfnIdentityPoolRoleAttachment,
  CfnUserPoolUser,
} from "aws-cdk-lib/aws-cognito";
import { Effect, FederatedPrincipal, PolicyDocument, PolicyStatement, Role } from "aws-cdk-lib/aws-iam";
import { Aws, ArnFormat, CfnResource, Duration, RemovalPolicy, Stack } from "aws-cdk-lib";
import { CfnPolicy } from "aws-cdk-lib/aws-iot";
import { Construct } from "constructs";

/**
 * CognitoAuthConstruct props
 *
 * @interface CognitoAuthConstructProps
 */
export interface CognitoAuthConstructProps {
  adminEmail: string;
  adminName: string;
  apiId: string;
  webAppURL: string;
  scenariosBucketArn: string;
}

export class CognitoAuthConstruct extends Construct {
  cognitoIdentityPoolId: string;
  cognitoUserPoolClientId: string;
  cognitoUserPoolId: string;
  public iotPolicy: CfnPolicy;

  constructor(scope: Construct, id: string, props: CognitoAuthConstructProps) {
    super(scope, id);

    const dltIotPolicy = new CfnPolicy(this, "IoT-Policy", {
      policyDocument: new PolicyDocument({
        statements: [
          new PolicyStatement({
            effect: Effect.ALLOW,
            actions: ["iot:Connect"],
            resources: [
              Stack.of(this).formatArn({
                service: "iot",
                resource: "client",
                resourceName: "*",
                arnFormat: ArnFormat.SLASH_RESOURCE_NAME,
              }),
            ],
          }),
          new PolicyStatement({
            effect: Effect.ALLOW,
            actions: ["iot:Subscribe"],
            resources: [
              Stack.of(this).formatArn({
                service: "iot",
                resource: "topicfilter",
                resourceName: "*",
                arnFormat: ArnFormat.SLASH_RESOURCE_NAME,
              }),
            ],
          }),
          new PolicyStatement({
            effect: Effect.ALLOW,
            actions: ["iot:Receive"],
            resources: [
              Stack.of(this).formatArn({
                service: "iot",
                resource: "topic",
                resourceName: "*",
                arnFormat: ArnFormat.SLASH_RESOURCE_NAME,
              }),
            ],
          }),
        ],
      }),
    });
    this.iotPolicy = dltIotPolicy;
    dltIotPolicy.addMetadata("cfn_nag", {
      rules_to_suppress: [
        {
          id: "W11",
          reason: "Cannot specify the resource to attach policy to identity",
        },
      ],
    });

    const cognitoUserPool = new UserPool(this, "DLTUserPool", {
      autoVerify: {
        email: true,
      },
      passwordPolicy: {
        minLength: 12,
        requireLowercase: true,
        requireDigits: true,
        requireSymbols: true,
        requireUppercase: true,
      },
      removalPolicy: RemovalPolicy.DESTROY,
      selfSignUpEnabled: false,
      signInAliases: {
        email: true,
        username: true,
      },
      standardAttributes: {
        email: {
          required: true,
        },
      },
      userInvitation: {
        emailSubject: "Welcome to Distributed Load Testing",
        emailBody: `
                <p>
                   Please use the credentials below to login to the Distributed Load Testing console.
                </p>
                <p>
                    Username: <strong>{username}</strong>
                </p>
                <p>
                    Password: <strong>{####}</strong>
                </p>
                <p>
                    Console: <strong>${props.webAppURL}/</strong>
                </p>
              `,
        smsMessage: "Your username is {username} and temporary password is {####}.",
      },
      userPoolName: `${Aws.STACK_NAME}-user-pool`,
    });
    (cognitoUserPool.node.defaultChild as CfnUserPool).userPoolAddOns = { advancedSecurityMode: "ENFORCED" };
    this.cognitoUserPoolId = cognitoUserPool.userPoolId;

    const clientWriteAttributes = new ClientAttributes().withStandardAttributes({
      address: true,
      email: true,
      phoneNumber: true,
    });

    const cognitoUserPoolClient = new UserPoolClient(this, "DLTUserPoolClient", {
      userPoolClientName: `${Aws.STACK_NAME}-userpool-client`,
      userPool: cognitoUserPool,
      generateSecret: false,
      writeAttributes: clientWriteAttributes,
      refreshTokenValidity: Duration.days(1),
    });

    this.cognitoUserPoolClientId = cognitoUserPoolClient.userPoolClientId;

    const cognitoIdentityPool = new CfnIdentityPool(this, "DLTIdentityPool", {
      allowUnauthenticatedIdentities: false,
      cognitoIdentityProviders: [
        {
          clientId: this.cognitoUserPoolClientId,
          providerName: cognitoUserPool.userPoolProviderName,
        },
      ],
    });

    this.cognitoIdentityPoolId = cognitoIdentityPool.ref;

    const apiProdExecuteArn = Stack.of(this).formatArn({
      service: "execute-api",
      resource: props.apiId,
      resourceName: "prod/*",
    });
    const cognitoAuthorizedRole = new Role(this, "DLTCognitoAuthorizedRole", {
      assumedBy: new FederatedPrincipal(
        "cognito-identity.amazonaws.com",
        {
          StringEquals: { "cognito-identity.amazonaws.com:aud": this.cognitoIdentityPoolId },
          "ForAnyValue:StringLike": { "cognito-identity.amazonaws.com:amr": "authenticated" },
        },
        "sts:AssumeRoleWithWebIdentity"
      ),
      description: `${Aws.STACK_NAME} Identity Pool authenticated role`,
      inlinePolicies: {
        InvokeApiPolicy: new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["execute-api:Invoke"],
              resources: [apiProdExecuteArn],
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["s3:PutObject", "s3:GetObject"],
              resources: [`${props.scenariosBucketArn}/public/*`, `${props.scenariosBucketArn}/cloudWatchImages/*`],
            }),
          ],
        }),
        IoTPolicy: new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["iot:AttachPrincipalPolicy"],
              resources: ["*"],
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["iot:Connect"],
              resources: [
                Stack.of(this).formatArn({
                  service: "iot",
                  resource: "client",
                  resourceName: "*",
                  arnFormat: ArnFormat.SLASH_RESOURCE_NAME,
                }),
              ],
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["iot:Subscribe"],
              resources: [
                Stack.of(this).formatArn({
                  service: "iot",
                  resource: "topicfilter",
                  resourceName: "*",
                  arnFormat: ArnFormat.SLASH_RESOURCE_NAME,
                }),
              ],
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["iot:Receive"],
              resources: [
                Stack.of(this).formatArn({
                  service: "iot",
                  resource: "topic",
                  resourceName: "*",
                  arnFormat: ArnFormat.SLASH_RESOURCE_NAME,
                }),
              ],
            }),
          ],
        }),
      },
    });
    (cognitoAuthorizedRole.node.defaultChild as CfnResource).addMetadata("cfn_nag", {
      rules_to_suppress: [
        {
          id: "W11",
          reason: "iot:AttachPrincipalPolicy does not allow for resource specification",
        },
      ],
    });

    const cognitoUnauthorizedRole = new Role(this, "DLTCognitoUnauthorizedRole", {
      assumedBy: new FederatedPrincipal(
        "cognito-identity.amazonaws.com",
        {
          StringEquals: { "cognito-identity.amazonaws.com:aud": this.cognitoIdentityPoolId },
          "ForAnyValue:StringLike": { "cognito-identity.amazonaws.com:amr": "unauthenticated" },
        },
        "sts:AssumeRoleWithWebIdentity"
      ),
    });

    new CfnIdentityPoolRoleAttachment(this, "CognitoAttachRole", {
      identityPoolId: this.cognitoIdentityPoolId,
      roles: {
        unauthenticated: cognitoUnauthorizedRole.roleArn,
        authenticated: cognitoAuthorizedRole.roleArn,
      },
    });

    new CfnUserPoolUser(this, "CognitoUser", {
      desiredDeliveryMediums: ["EMAIL"],
      forceAliasCreation: true,
      userAttributes: [
        { name: "email", value: props.adminEmail },
        { name: "nickname", value: props.adminName },
        { name: "email_verified", value: "true" },
      ],
      username: props.adminName,
      userPoolId: this.cognitoUserPoolId,
    });
  }
}
