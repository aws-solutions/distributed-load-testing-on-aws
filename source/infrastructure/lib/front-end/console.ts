// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { CloudFrontToS3 } from "@aws-solutions-constructs/aws-cloudfront-s3";
import { Bucket, IBucket } from "aws-cdk-lib/aws-s3";
import { Construct } from "constructs";
import { BucketDeployment, Source as S3Source } from "aws-cdk-lib/aws-s3-deployment";
import * as path from "path";

export interface DLTConsoleConstructProps {
  readonly s3LogsBucket: Bucket;
  readonly solutionId: string;
}

export class DLTConsoleConstruct extends Construct {
  public webAppURL: string;
  public consoleBucketArn: string;
  public consoleBucket: IBucket;

  constructor(scope: Construct, id: string, props: DLTConsoleConstructProps) {
    super(scope, id);

    // Create the standard region CloudFront distribution
    const dltS3CloudFrontDist = new CloudFrontToS3(this, "DLTDashboardS3", {
      bucketProps: {
        serverAccessLogsBucket: props.s3LogsBucket,
        serverAccessLogsPrefix: "console-bucket-access/",
      },
      cloudFrontDistributionProps: {
        comment: "Website distribution for the Distributed Load Testing solution",
        enableLogging: true,
        errorResponses: [
          { httpStatus: 403, responseHttpStatus: 200, responsePagePath: "/index.html" },
          { httpStatus: 404, responseHttpStatus: 200, responsePagePath: "/index.html" },
        ],
        httpVersion: "http2",
        logBucket: props.s3LogsBucket,
        logFilePrefix: "cloudfront-logs/",
      },
      insertHttpSecurityHeaders: false,
    });

    this.consoleBucket = dltS3CloudFrontDist.s3BucketInterface;
    this.webAppURL = `https://${dltS3CloudFrontDist.cloudFrontWebDistribution.domainName}`;
    this.consoleBucketArn = dltS3CloudFrontDist.s3BucketInterface.bucketArn;

    new BucketDeployment(this, "DeployWebsite", {
      sources: [S3Source.asset(path.join(__dirname, "../../../console/build"))], // build react app before cdk deploy
      destinationBucket: this.consoleBucket,
      exclude: ["assets/aws_config.js"],
    });
  }
}
