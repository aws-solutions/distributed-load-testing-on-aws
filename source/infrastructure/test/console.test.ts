// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { App, DefaultStackSynthesizer, Stack, CfnCondition, Aws, Fn } from "aws-cdk-lib";
import { DLTConsoleConstruct } from "../lib/front-end/console";
import { Bucket } from "aws-cdk-lib/aws-s3";
import { createTemplateWithoutS3Key } from "./snapshot_helpers";

test("DLT API Test", () => {
  const app = new App();
  const stack = new Stack(app, "DLTStack", {
    synthesizer: new DefaultStackSynthesizer({
      generateBootstrapVersionRule: false,
    }),
  });
  const testSourceBucket = new Bucket(stack, "testSourceCodeBucket");

  const console = new DLTConsoleConstruct(stack, "TestConsoleResources", {
    s3LogsBucket: testSourceBucket,
    solutionId: "SO0062",
  });

  expect(createTemplateWithoutS3Key(stack)).toMatchSnapshot();
  expect(console.webAppURL).toBeDefined();
  expect(console.consoleBucket).toBeDefined();
  expect(console.consoleBucketArn).toBeDefined();
});
