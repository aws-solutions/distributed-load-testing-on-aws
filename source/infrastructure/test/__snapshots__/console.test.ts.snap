// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DLT API Test 1`] = `
{
  "Resources": {
    "TestConsoleResourcesDLTCloudFrontToS3CloudFrontDistribution4534927E": {
      "Metadata": {
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W70",
              "reason": "Since the distribution uses the CloudFront domain name, CloudFront automatically sets the security policy to TLSv1 regardless of the value of MinimumProtocolVersion",
            },
          ],
        },
      },
      "Properties": {
        "DistributionConfig": {
          "Comment": "Website distribution for the Distributed Load Testing solution",
          "CustomErrorResponses": [
            {
              "ErrorCode": 403,
              "ResponseCode": 200,
              "ResponsePagePath": "/index.html",
            },
            {
              "ErrorCode": 404,
              "ResponseCode": 200,
              "ResponsePagePath": "/index.html",
            },
          ],
          "DefaultCacheBehavior": {
            "CachePolicyId": "658327ea-f89d-4fab-a63d-7e88639e58f6",
            "Compress": true,
            "TargetOriginId": "DLTStackTestConsoleResourcesDLTCloudFrontToS3CloudFrontDistributionOrigin1AA2FCD06",
            "ViewerProtocolPolicy": "redirect-to-https",
          },
          "DefaultRootObject": "index.html",
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Logging": {
            "Bucket": {
              "Fn::GetAtt": [
                "testSourceCodeBucketC577B176",
                "RegionalDomainName",
              ],
            },
            "Prefix": "cloudfront-logs/",
          },
          "Origins": [
            {
              "DomainName": {
                "Fn::GetAtt": [
                  "TestConsoleResourcesDLTCloudFrontToS3S3BucketAC4D9E17",
                  "RegionalDomainName",
                ],
              },
              "Id": "DLTStackTestConsoleResourcesDLTCloudFrontToS3CloudFrontDistributionOrigin1AA2FCD06",
              "S3OriginConfig": {
                "OriginAccessIdentity": {
                  "Fn::Join": [
                    "",
                    [
                      "origin-access-identity/cloudfront/",
                      {
                        "Ref": "TestConsoleResourcesDLTCloudFrontToS3CloudFrontDistributionOrigin1S3OriginDB11D6D3",
                      },
                    ],
                  ],
                },
              },
            },
          ],
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "TestConsoleResourcesDLTCloudFrontToS3CloudFrontDistributionOrigin1S3OriginDB11D6D3": {
      "Properties": {
        "CloudFrontOriginAccessIdentityConfig": {
          "Comment": "Identity for DLTStackTestConsoleResourcesDLTCloudFrontToS3CloudFrontDistributionOrigin1AA2FCD06",
        },
      },
      "Type": "AWS::CloudFront::CloudFrontOriginAccessIdentity",
    },
    "TestConsoleResourcesDLTCloudFrontToS3S3BucketAC4D9E17": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "BucketEncryption": {
          "ServerSideEncryptionConfiguration": [
            {
              "ServerSideEncryptionByDefault": {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "LifecycleConfiguration": {
          "Rules": [
            {
              "NoncurrentVersionTransitions": [
                {
                  "StorageClass": "GLACIER",
                  "TransitionInDays": 90,
                },
              ],
              "Status": "Enabled",
            },
          ],
        },
        "LoggingConfiguration": {
          "DestinationBucketName": {
            "Ref": "testSourceCodeBucketC577B176",
          },
          "LogFilePrefix": "console-bucket-access/",
        },
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "VersioningConfiguration": {
          "Status": "Enabled",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "TestConsoleResourcesDLTCloudFrontToS3S3BucketPolicy2004C0F3": {
      "Metadata": {
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "F16",
              "reason": "Public website bucket policy requires a wildcard principal",
            },
          ],
        },
      },
      "Properties": {
        "Bucket": {
          "Ref": "TestConsoleResourcesDLTCloudFrontToS3S3BucketAC4D9E17",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:*",
              "Condition": {
                "Bool": {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": {
                "AWS": "*",
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "TestConsoleResourcesDLTCloudFrontToS3S3BucketAC4D9E17",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "TestConsoleResourcesDLTCloudFrontToS3S3BucketAC4D9E17",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": {
                "CanonicalUser": {
                  "Fn::GetAtt": [
                    "TestConsoleResourcesDLTCloudFrontToS3CloudFrontDistributionOrigin1S3OriginDB11D6D3",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "TestConsoleResourcesDLTCloudFrontToS3S3BucketAC4D9E17",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "testSourceCodeBucketC577B176": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AccessControl": "LogDeliveryWrite",
        "OwnershipControls": {
          "Rules": [
            {
              "ObjectOwnership": "ObjectWriter",
            },
          ],
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
